drv/stm32h7-usart/src/lib.rs

```diff
@@ -0,0 +1,139 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! USART interface.
+//!
+//! This was formerly a driver and the implementation still looks largely
+//! driver-like in that it interacts with other drivers and system interrupts,
+//! but USARTs are inherently single-owner. Structuring USARTs as a lib allows
+//! the calling task to use the USART more directly than if it were a
+//! full-fledged driver of its own.
+
+#![no_std]
+
+pub use drv_stm32xx_sys_api;
+
+#[cfg(feature = "h743")]
+pub use stm32h7::stm32h743 as device;
+
+#[cfg(feature = "h753")]
+pub use stm32h7::stm32h753 as device;
+
+use drv_stm32xx_sys_api::{Alternate, Peripheral, PinSet, Sys};
+use unwrap_lite::UnwrapLite;
+
+/// Handle to an enabled USART device.
+pub struct Usart {
+    usart: &'static device::usart1::RegisterBlock,
+}
+
+impl Usart {
+    /// Turn on the `USART` described by `usart`, `peripheral`, `tx_rx_mask`,
+    /// and `alternate`, with the baud rate defined by `clock_hz` and
+    /// `baud_rate`.
+    ///
+    /// Enables interrupts from the `USART` when bytes are available to receive;
+    /// the caller is responsible for enabling and handling the corresponding
+    /// kernel interrupt.
+    pub fn turn_on(
+        sys: &Sys,
+        usart: &'static device::usart1::RegisterBlock,
```

Comment:
As far as I can find (in the code in `drv/` at least) `usart1::RegisterBlock` is just a pointer. Where are all the fields and methods coming from?

Reply:
Linking to the stm32h7 crate is annoyingly difficult - the docs.rs build is broken, and the repo doesn't have code in it since it's autogenerated. `usart1::RegisterBlock` is a normal Rust struct; the `USART3::ptr()` method gives us a pointer to it, but we then convert that to a `&'static ref`, at which point we can use it like any other ref-to-a-struct. It has fields for all the relevant USART registers:

```rust
#[repr(C)]
pub struct RegisterBlock {
    pub cr1: Reg<CR1_SPEC>,
    pub cr2: Reg<CR2_SPEC>,
    pub cr3: Reg<CR3_SPEC>,
    pub brr: Reg<BRR_SPEC>,
    pub gtpr: Reg<GTPR_SPEC>,
    pub rtor: Reg<RTOR_SPEC>,
    pub rqr: Reg<RQR_SPEC>,
    pub isr: Reg<ISR_SPEC>,
    pub icr: Reg<ICR_SPEC>,
    pub rdr: Reg<RDR_SPEC>,
    pub tdr: Reg<TDR_SPEC>,
    pub presc: Reg<PRESC_SPEC>,
}
```

Reply:
I found the [stm32h7-nightly](https://github.com/stm32-rs/stm32-rs-nightlies) builds helpful here. I got lost trying to sort out some quirks in stm32h7 related to the RNG and having the generated code (not just the SVDs) was bigly helpful.